src/cloud-providers/gcp/provider.go

```diff
@@ -0,0 +1,191 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package gcp
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"log"
+	"net/netip"
+
+	compute "cloud.google.com/go/compute/apiv1"
+	computepb "cloud.google.com/go/compute/apiv1/computepb"
+	provider "github.com/confidential-containers/cloud-api-adaptor/src/cloud-providers"
+	"github.com/confidential-containers/cloud-api-adaptor/src/cloud-providers/util"
+	"github.com/confidential-containers/cloud-api-adaptor/src/cloud-providers/util/cloudinit"
+	"golang.org/x/oauth2/google"
+	option "google.golang.org/api/option"
+	proto "google.golang.org/protobuf/proto"
+)
+
+var logger = log.New(log.Writer(), "[adaptor/cloud/gcp] ", log.LstdFlags|log.Lmsgprefix)
+var computeScope = "https://www.googleapis.com/auth/compute"
+
+const maxInstanceNameLen = 63
+
+type gcpProvider struct {
+	serviceConfig   *Config
+	instancesClient *compute.InstancesClient
+}
+
+func (p *gcpProvider) ConfigVerifier() error {
+	return nil
+}
+
+func NewProvider(config *Config) (provider.Provider, error) {
+	logger.Printf("gcp config: %#v", config.Redact())
+	provider := &gcpProvider{
+		serviceConfig:   config,
+		instancesClient: nil,
+	}
+	if config.GcpCredentials != "" {
+		creds, err := google.CredentialsFromJSON(context.TODO(), []byte(config.GcpCredentials), computeScope)
+		if err != nil {
+			return nil, fmt.Errorf("configuration error when using creds: %s", err)
+		}
+		provider.instancesClient, err = compute.NewInstancesRESTClient(context.TODO(), option.WithCredentials(creds))
+		if err != nil {
+			return nil, fmt.Errorf("NewInstancesRESTClient with credentials error: %s", err)
+		}
+	} else {
+		var err error
+		provider.instancesClient, err = compute.NewInstancesRESTClient(context.TODO())
+		if err != nil {
+			return nil, fmt.Errorf("NewInstancesRESTClient error: %s", err)
+		}
+	}
+	return provider, nil
+}
+
+func getIPs(instance *computepb.Instance) ([]netip.Addr, error) {
+	var podNodeIPs []netip.Addr
+	for _, nic := range instance.GetNetworkInterfaces() {
+		for _, access := range nic.GetAccessConfigs() {
+			ipStr := access.GetNatIP()
+			ip, err := netip.ParseAddr(ipStr)
+			if err != nil {
+				return nil, fmt.Errorf("failed to parse pod node IP %q: %w", ipStr, err)
+			}
+			podNodeIPs = append(podNodeIPs, ip)
+			logger.Printf("%s", ip.String())
+		}
+	}
+	return podNodeIPs, nil
+}
+
+func (p *gcpProvider) CreateInstance(ctx context.Context, podName, sandboxID string, cloudConfig cloudinit.CloudConfigGenerator, spec provider.InstanceTypeSpec) (*provider.Instance, error) {
+
+	instanceName := util.GenerateInstanceName(podName, sandboxID, maxInstanceNameLen)
+	logger.Printf("CreateInstance: name: %q", instanceName)
+
+	userData, err := cloudConfig.Generate()
+	if err != nil {
+		return nil, err
+	}
+
+	//Convert userData to base64
+	userDataEnc := base64.StdEncoding.EncodeToString([]byte(userData))
+	logger.Printf("userDataEnc:  %s", userDataEnc)
+
+	insertReq := &computepb.InsertInstanceRequest{
```

Comment:
You will need to improve the InstanceRequest to allow users to customize the VM. One concept that almost all providers implemented is of "flexible instance sizes" (please see https://github.com/confidential-containers/cloud-api-adaptor/issues/321).

Enable/disable confidential VM is also something desirable that I mentioned elsewhere.

Reply:
I would add this to the list, too. If you don't mind.